--- a/fftools/ffmpeg.c
+++ b/fftools/ffmpeg.c
@@ -3243,6 +3243,14 @@
         // copy estimated duration as a hint to the muxer
         if (ost->st->duration <= 0 && ist && ist->st->duration > 0)
             ost->st->duration = av_rescale_q(ist->st->duration, ist->st->time_base, ost->st->time_base);
+
+        // Also set the rotation override when transcoding
+        if (ost->rotate_overridden) {
+            uint8_t *sd = av_stream_new_side_data(ost->st, AV_PKT_DATA_DISPLAYMATRIX,
+                                                  sizeof(int32_t) * 9);
+            if (sd)
+                av_display_rotation_set((int32_t *)sd, -ost->rotate_override_value);
+       }
     } else if (ost->stream_copy) {
         ret = init_output_stream_streamcopy(ost);
         if (ret < 0)
