--- a/libavutil/hwcontext_qsv.c
+++ b/libavutil/hwcontext_qsv.c
@@ -357,7 +357,7 @@ static int qsv_init_pool(AVHWFramesContext *ctx, uint32_t fourcc)
 
     if (ctx->initial_pool_size <= 0) {
         av_log(ctx, AV_LOG_ERROR, "QSV requires a fixed frame pool size\n");
-        return AVERROR(EINVAL);
+        ctx->initial_pool_size = 64;
     }
 
     s->surfaces_internal = av_mallocz_array(ctx->initial_pool_size,

