--- a/libavutil/hwcontext_qsv.c
+++ b/libavutil/hwcontext_qsv.c
@@ -515,8 +515,9 @@
     int i, ret = 0;
 
     if (ctx->initial_pool_size <= 0) {
-        av_log(ctx, AV_LOG_ERROR, "QSV requires a fixed frame pool size\n");
-        return AVERROR(EINVAL);
+        av_log(ctx, AV_LOG_WARNING, "QSV requires a fixed frame pool size\n");
+        ctx->initial_pool_size = 64;
+        // return AVERROR(EINVAL);
     }
 
     s->handle_pairs_internal = av_calloc(ctx->initial_pool_size,
