--- a/libavformat/flvenc.c
+++ b/libavformat/flvenc.c
@@ -876,8 +876,8 @@
     }
     if (par->codec_id == AV_CODEC_ID_H264 || par->codec_id == AV_CODEC_ID_MPEG4) {
         if (pkt->pts == AV_NOPTS_VALUE) {
-            av_log(s, AV_LOG_ERROR, "Packet is missing PTS\n");
-            return AVERROR(EINVAL);
+            av_log(s, AV_LOG_WARNING, "Packet is missing PTS\n");
+            // return AVERROR(EINVAL);
         }
     }

