diff -Naur ffmpeg-2.1/libavformat/mov.c ffmpeg-2.1-patched/libavformat/mov.c
--- ffmpeg-2.1/libavformat/mov.c	2014-01-08 16:12:09.000000000 +0100
+++ ffmpeg-2.1-patched/libavformat/mov.c	2014-01-08 16:11:59.000000000 +0100
@@ -1779,7 +1779,7 @@
     if (!entries)
     {
         sc->keyframe_absent = 1;
-        if (!st->need_parsing)
+        if (!st->need_parsing && st->codec->codec_type == AVMEDIA_TYPE_VIDEO)
             st->need_parsing = AVSTREAM_PARSE_HEADERS;
         return 0;
     }
