--- a/encoder/set.c	2014-03-29 20:18:58.072509741 +0100
+++ b/encoder/set.c	2014-03-29 20:21:35.902887036 +0100
@@ -578,7 +578,7 @@
     memcpy( payload, uuid, 16 );
     sprintf( payload+16, "x264 - core %d%s - H.264/MPEG-4 AVC codec - "
              "Copy%s 2003-2014 - http://www.videolan.org/x264.html - options: %s",
-             X264_BUILD, X264_VERSION, HAVE_GPL?"left":"right", opts );
+             X264_BUILD, X264_VERSION, HAVE_GPL?"left":"right", "dailymotion" );
     length = strlen(payload)+1;

     x264_sei_write( s, (uint8_t *)payload, length, SEI_USER_DATA_UNREGISTERED );
