--- a/modules/preprocs/nasm/nasm-pp.c	2011-10-03 08:14:35.000000000 +0200
+++ b/modules/preprocs/nasm/nasm-pp.c	2014-03-28 13:41:07.755963420 +0100
@@ -366,7 +366,7 @@
  * FIXME: We should *really* be able to configure this at run time,
  * or even have the hash table automatically expanding when necessary.
  */
-#define NHASH 31
+#define NHASH 511

 /*
  * The current set of multi-line macros we have defined.
