--- a/builder/FindFunctions.cmake
+++ b/builder/FindFunctions.cmake
@@ -116,7 +116,7 @@
   get_property( PLUGINS_CFG GLOBAL PROPERTY PROP_PLUGINS_CFG )
   file(WRITE ${directory}/plugins.cfg ${PLUGINS_CFG})

-  install( FILES ${directory}/plugins.cfg DESTINATION ${MFX_PLUGINS_CONF_DIR} )
+  install( FILES ${directory}/plugins.cfg DESTINATION ${CMAKE_INSTALL_FULL_DATADIR}/mfx )

 endfunction()

--- a/_studio/mfx_lib/CMakeLists.txt
--- b/_studio/mfx_lib/CMakeLists.txt
@@ -224,7 +224,7 @@
 set_target_properties(${mfxlibname} PROPERTIES   VERSION ${mfx_version_major}.${mfx_version_minor})
 set_target_properties(${mfxlibname} PROPERTIES SOVERSION ${mfx_version_major})

-install(TARGETS ${mfxlibname} LIBRARY DESTINATION ${MFX_MODULES_DIR})
+install(TARGETS ${mfxlibname} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR})

 set( PKG_CONFIG_FNAME "${CMAKE_LIB_DIR}/${CMAKE_BUILD_TYPE}/lib${mfxlibname}.pc")
 configure_file("${CMAKE_CURRENT_SOURCE_DIR}/pkg-config.pc.cmake" ${PKG_CONFIG_FNAME} @ONLY)
--- a/_studio/hevce_hw/CMakeLists.txt
--- b/_studio/hevce_hw/CMakeLists.txt
@@ -53,4 +53,4 @@
 set( defs "${defs} ${version_defs}" )
 gen_plugins_cfg( "HEVC_Encoder_HW" ${HEVC_Encoder_HW_GUID} ${plugin_name} "02" "HEVC" )
 make_library( ${plugin_name} hw shared )
-install( TARGETS ${plugin_name} LIBRARY DESTINATION ${MFX_PLUGINS_DIR} )
+install( TARGETS ${plugin_name} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/mfx )
--- a/_studio/hevc_fei/CMakeLists.txt
--- b/_studio/hevc_fei/CMakeLists.txt
@@ -49,5 +49,5 @@
   set( defs "${defs} ${version_defs}" )
   gen_plugins_cfg("HEVC_Encoder_HW" ${HEVC_FEI_Encoder_HW_GUID} ${plugin_name} "02" "HEVC")
   make_library( ${plugin_name} hw shared )
-  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${MFX_PLUGINS_DIR} )
+  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/mfx )
 endif( )
--- a/_studio/vp9e_hw/CMakeLists.txt
--- b/_studio/vp9e_hw/CMakeLists.txt
@@ -47,4 +47,4 @@
 set( defs "${defs} ${version_defs}" )
 gen_plugins_cfg( "VP9_Encoder_HW" ${VP9_Encoder_HW_GUID} ${plugin_name} "02" "VP9" )
 make_library( ${plugin_name} hw shared )
-install( TARGETS ${plugin_name} LIBRARY DESTINATION ${MFX_PLUGINS_DIR} )
+install( TARGETS ${plugin_name} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/mfx )
--- a/_studio/mfx_lib/plugin/CMakeLists.txt
--- b/_studio/mfx_lib/plugin/CMakeLists.txt
@@ -190,7 +190,7 @@
   set( defs "${defs} ${version_defs}" )
   gen_plugins_cfg("H264la_Encoder" ${H264LA_ENCODER_GUID} ${plugin_name} "04" "AVC")
   make_library( ${plugin_name} hw shared)
-  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${MFX_PLUGINS_DIR} )
+  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/mfx )
 endif()

 ###############################################################################
@@ -229,7 +229,7 @@
   set( defs "${defs} ${version_defs}" )
   gen_plugins_cfg("HEVC_Decoder_HW" ${HEVC_DECODER_HW_GUID} ${plugin_name} "01" "HEVC")
   make_library( ${plugin_name} hw shared)
-  install( TARGETS ${plugin_name} ${plugin_name} LIBRARY DESTINATION ${MFX_PLUGINS_DIR} )
+  install( TARGETS ${plugin_name} ${plugin_name} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/mfx )
 endif()

 ###############################################################################
@@ -269,7 +269,7 @@
   set( defs "${defs} ${version_defs}" )
   gen_plugins_cfg( "VP8_Decoder_HW" ${VP8_DECODER_HW_GUID} ${plugin_name} "01" "VP8" )
   make_library( ${plugin_name} hw shared)
-  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${MFX_PLUGINS_DIR} )
+  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/mfx )
 endif()

 # =============================================================================
@@ -311,5 +311,5 @@
   set( defs "${defs} ${version_defs}" )
   gen_plugins_cfg( "VP9_Decoder_HW" ${VP9_DECODER_HW_GUID} ${plugin_name} "01" "VP9" )
   make_library( ${plugin_name} hw shared)
-  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${MFX_PLUGINS_DIR} )
+  install( TARGETS ${plugin_name} LIBRARY DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/mfx )
 endif()
